rust/agama-lib/src/storage/client/zfcp.rs

```diff
@@ -0,0 +1,207 @@
+                match ZFCPDisk::try_from(properties) {
+                    Ok(device) => {
+                        devices.push((path, device));
+                    }
+                    Err(error) => {
+                        log::warn!("Not a valid zFCP disk: {}", error);
+                    }
+                }
+            }
+        }
+        Ok(devices)
+    }
+
+    pub async fn get_controllers(
+        &self,
+    ) -> Result<Vec<(OwnedObjectPath, ZFCPController)>, ServiceError> {
+        let managed_objects = self.object_manager_proxy.get_managed_objects().await?;
+
+        let mut devices: Vec<(OwnedObjectPath, ZFCPController)> = vec![];
+        for (path, ifaces) in managed_objects {
+            if let Some(properties) = ifaces.get("org.opensuse.Agama.Storage1.ZFCP.Controller") {
+                // TODO: maybe move it to model? but it needs also additional calls to dbus
+                let path_s = Self::controller_id_from_path(&path);
+                devices.push((
+                    path,
+                    ZFCPController {
+                        id: path_s.clone(),
+                        channel: get_property(properties, "Channel")?,
+                        lun_scan: get_property(properties, "LUNScan")?,
+                        active: get_property(properties, "Active")?,
+                        luns_map: self.get_luns_map(path_s.as_str()).await?,
+                    },
+                ))
+            }
+        }
+        Ok(devices)
+    }
+
+    /// Gets controller id that can be used in other controller methods from its dbus path
+    ///
+    /// common case:
+    /// ```
+    ///     let path = zbus::zvariant::OwnedObjectPath::try_from("/test/controllers/2").unwrap();
+    ///     let id = agama_lib::storage::client::zfcp::ZFCPClient::controller_id_from_path(&path);
+    ///     assert_eq!(id, "2".to_string());
+    /// ```
+    ///
+    /// edge case with malformed path that will panic in Owned path already
+    /// ```should_panic
+    ///     let path = zbus::zvariant::OwnedObjectPath::try_from("mangled").unwrap();
+    ///     let id = agama_lib::storage::client::zfcp::ZFCPClient::controller_id_from_path(&path);
+    ///     assert_eq!(id, "mangled".to_string());
+    /// ```
+    pub fn controller_id_from_path(path: &OwnedObjectPath) -> String {
+        let mut path_s = path.to_string();
+        let slash_pos = path_s.rfind("/").unwrap_or(0);
+        path_s.drain(..slash_pos + 1);
+        path_s
+    }
+
+    async fn get_controller_proxy(
+        &self,
+        controller_id: &str,
+    ) -> Result<ZFCPControllerProxy, ServiceError> {
+        let dbus = ZFCPControllerProxy::builder(&self.connection)
+            .path(ZFCP_CONTROLLER_PREFIX.to_string() + "/" + controller_id)?
+            .build()
+            .await?;
+        Ok(dbus)
+    }
+
+    // TODO: does all those method with controller id as parameter deserve to be moved to controller model?
```

Comment:
np: Do you plan to do something with this comment? Otherwise, I would remove the `TODO` prefix.

Reply:
@jreidinger I guess it is a question you wanted to raise and discuss, so, will remove the comment but maybe this thread could be used for having that discussion... by now it looks good to me to leave the definition where it is, @imobachgs, what do you think?

Reply:
It is fine. I do not like that we use two different approaches, but let's move on.